index-scheduler/src/lib.rs

```diff
@@ -378,37 +421,93 @@ impl IndexScheduler {
+                        include_processing_tasks = true;
+                        status_tasks |= &tasks_processing;
+                    }
+                    status => status_tasks |= &self.get_status(&rtxn, *status)?,
+                };
+            }
+            if !include_processing_tasks {
+                tasks -= &tasks_processing;
             }
             tasks &= status_tasks;
         }
 
+        if let Some(uids) = &query.uid {
+            let uids = RoaringBitmap::from_iter(uids);
+            tasks &= &uids;
+        }
+
         if let Some(kind) = &query.kind {
             let mut kind_tasks = RoaringBitmap::new();
             for kind in kind {
                 kind_tasks |= self.get_kind(&rtxn, *kind)?;
             }
-            tasks &= kind_tasks;
+            tasks &= &kind_tasks;
         }
 
         if let Some(index) = &query.index_uid {
             let mut index_tasks = RoaringBitmap::new();
             for index in index {
                 index_tasks |= self.index_tasks(&rtxn, index)?;
             }
-            tasks &= index_tasks;
+            tasks &= &index_tasks;
         }
 
+        // For the started_at filter, we need to treat the part of the tasks that are processing from the part of the
+        // tasks that are not processing. The non-processing ones are filtered normally while the processing ones
+        // are entirely removed unless the in-memory startedAt variable falls within the date filter.
+        // Once we have filtered the two subsets, we put them back together and assign it back to `tasks`.
+        tasks = {
+            let (mut filtered_non_processing_tasks, mut filtered_processing_tasks) =
+                (&tasks - &tasks_processing, &tasks & &tasks_processing);
+
+            // special case for Processing tasks
+            // in a loop because I want to break early if both query.after_started_at and query.before_started_at are None
+            // it doesn't actually loop
+            'block: loop {
+                let bounds = match (query.after_started_at, query.before_started_at) {
+                    (None, None) => break 'block,
+                    (None, Some(before)) => (Bound::Unbounded, Bound::Excluded(before)),
+                    (Some(after), None) => (Bound::Excluded(after), Bound::Unbounded),
+                    (Some(after), Some(before)) => {
+                        (Bound::Excluded(after), Bound::Excluded(before))
+                    }
+                };
+                let start = map_bound(bounds.0, |b| b.unix_timestamp_nanos());
+                let end = map_bound(bounds.1, |b| b.unix_timestamp_nanos());
+                let is_within_dates = RangeBounds::contains(
+                    &(start, end),
+                    &started_at_processing.unix_timestamp_nanos(),
+                );
+                if !is_within_dates {
+                    filtered_processing_tasks.clear();
+                }
```

Comment:
Instead of using a loop trick to be able to early return, why not putting that in a closure and calling it for the three cases? Or why not using an optional instead?

```
let x = match y {
    good => Some(y),
    bad => None,
};

if let Some(y) = x {
    // I do the thing here
}
```